From f2498499ca6c28b8fb1244dd764825bc75069944 Mon Sep 17 00:00:00 2001
From: Joel Hockey <joelhockey@chromium.org>
Date: Wed, 23 Jan 2019 00:32:07 -0800
Subject: [PATCH 1/7] lxd: Send metadata in CreateImage error response

Include fingerprint and size metadata in CreateImage error
response when image already exists.  When creating an image
from a container snapshot, it is useful to know the fingerprint
of an existing image.  It can currently be parsed from the error
string, but this makes it more explicit.

Signed-off-by: Joel Hockey <joelhockey@chromium.org>
---
 lxd/images.go | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/lxd/images.go b/lxd/images.go
index 1d2e863c..06acc335 100644
--- a/lxd/images.go
+++ b/lxd/images.go
@@ -247,7 +247,7 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, builddir st
 
 	_, _, err = d.cluster.ImageGet(info.Fingerprint, false, true)
 	if err == nil {
-		return nil, fmt.Errorf("The image already exists: %s", info.Fingerprint)
+		return &info, fmt.Errorf("The image already exists: %s", info.Fingerprint)
 	}
 
 	/* rename the the file to the expected name so our caller can use it */
@@ -696,6 +696,13 @@ func imagesPost(d *Daemon, r *http.Request) Response {
 				imagePublishLock.Unlock()
 			}
 		}
+		// Set the metadata if possible, even if there is an error
+		if info != nil {
+			metadata := make(map[string]string)
+			metadata["fingerprint"] = info.Fingerprint
+			metadata["size"] = strconv.FormatInt(info.Size, 10)
+			op.UpdateMetadata(metadata)
+		}
 		if err != nil {
 			return err
 		}
@@ -718,11 +725,6 @@ func imagesPost(d *Daemon, r *http.Request) Response {
 			}
 		}
 
-		// Set the metadata
-		metadata := make(map[string]string)
-		metadata["fingerprint"] = info.Fingerprint
-		metadata["size"] = strconv.FormatInt(info.Size, 10)
-		op.UpdateMetadata(metadata)
 		return nil
 	}
 
-- 
2.20.1.611.gfbb209baf1-goog

